runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/ActiveWorkState.java

```diff
@@ -0,0 +1,304 @@
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.collect.ImmutableList;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.collect.ImmutableMap;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Manages the active {@link Work} queues for their {@link ShardedKey}(s). Gives an interface to
+ * activate, queue, and complete {@link Work} (including invalidating stuck {@link Work}).
+ */
+@ThreadSafe
+final class ActiveWorkState {
+  private static final Logger LOG = LoggerFactory.getLogger(ActiveWorkState.class);
+
+  /**
+   * Map from {@link ShardedKey} to {@link Work} for the key. The first item in the {@link
+   * Queue<Work>} is actively processing.
+   */
+  @GuardedBy("this")
+  private final Map<ShardedKey, Deque<Work>> activeWork;
+
+  @GuardedBy("this")
+  private final WindmillStateCache.ForComputation computationStateCache;
+
+  private ActiveWorkState(
+      Map<ShardedKey, Deque<Work>> activeWork,
+      WindmillStateCache.ForComputation computationStateCache) {
+    this.activeWork = activeWork;
+    this.computationStateCache = computationStateCache;
+  }
+
+  static ActiveWorkState create(WindmillStateCache.ForComputation computationStateCache) {
+    return new ActiveWorkState(new HashMap<>(), computationStateCache);
+  }
+
+  @VisibleForTesting
+  static ActiveWorkState forTesting(
+      Map<ShardedKey, Deque<Work>> activeWork,
+      WindmillStateCache.ForComputation computationStateCache) {
+    return new ActiveWorkState(activeWork, computationStateCache);
+  }
+
+  /**
+   * Activates {@link Work} for the {@link ShardedKey}. Outcome can be 1 of 3 {@link
+   * ActivateWorkResult}
+   *
+   * <p>1. EXECUTE: The {@link ShardedKey} has not been seen before, create a {@link Queue<Work>}
+   * for the key. The caller should execute the work.
+   *
+   * <p>2. DUPLICATE: A work queue for the {@link ShardedKey} exists, and the work already exists in
+   * the {@link ShardedKey}'s work queue, mark the {@link Work} as a duplicate.
+   *
+   * <p>3. QUEUED: A work queue for the {@link ShardedKey} exists, and the work is not in the key's
+   * work queue, queue the work for later processing.
+   */
+  synchronized ActivateWorkResult activateWorkForKey(ShardedKey shardedKey, Work work) {
+    Deque<Work> workQueue = activeWork.getOrDefault(shardedKey, new ArrayDeque<>());
+
+    // This key does not have any work queued up on it. Create one, insert Work, and mark the work
+    // to be executed.
+    if (!activeWork.containsKey(shardedKey) || workQueue.isEmpty()) {
+      workQueue.addLast(work);
+      activeWork.put(shardedKey, workQueue);
+      return ActivateWorkResult.EXECUTE;
+    }
+
+    // Ensure we don't already have this work token queued.
+    for (Work queuedWork : workQueue) {
+      if (queuedWork.getWorkItem().getWorkToken() == work.getWorkItem().getWorkToken()) {
```

Comment:
if the cache token is different then it isn't a pure duplicate it could be a retry, in that case it might be better to take the more recent observed item (guessing it is more likely the newer one) or keep both.

Reply:
sgtm, I don't think this is currently done.

do we want to queue the retry then? What happens if we process both, would Windmill make sure only 1 is committed w/ the exactly once processing?

in more detail if we take the most recent observed item, then would we have to remove the older Work(sameWorkToken, existingCacheToken) from the queue, and add the more recent or passed in Work(sameWorkToken, newCacheToken) to the queue?

Reply:
discussed offline, will follow up in next cl